From 9fae298d930ad6323e180a57c3b55242e6a94a58 Mon Sep 17 00:00:00 2001
From: "Dr. Greg Wettstein" <greg@enjellic.com>
Date: Thu, 2 Feb 2017 16:13:20 -0600
Subject: [PATCH 07/24] Implement actor/subject template relationship.

In order to simplify the implementation of the measurement of the
actor/subject event the iso-identity template was modified to
explicitly call out actor and subject fields.

The actor field implements identical behavior to the 'id' field
which it replaces.  The subject field implementation at this
point is largely invariant to the 'd-ng' field which it replaces.

This commit is essentially a staging commit in order to begin
refining the implementation of a measurement over all of the
attributes which will be considered in a subject field.
---
 security/integrity/ima/ima_template.c     |  6 ++++--
 security/integrity/ima/ima_template_lib.c | 23 ++++++++++++++++++++---
 security/integrity/ima/ima_template_lib.h |  6 ++++--
 3 files changed, 28 insertions(+), 7 deletions(-)

diff --git a/security/integrity/ima/ima_template.c b/security/integrity/ima/ima_template.c
index 1011317c3eab..4ef93ec89993 100644
--- a/security/integrity/ima/ima_template.c
+++ b/security/integrity/ima/ima_template.c
@@ -24,7 +24,7 @@ static struct ima_template_desc defined_templates[] = {
 	{.name = IMA_TEMPLATE_IMA_NAME, .fmt = IMA_TEMPLATE_IMA_FMT},
 	{.name = "ima-ng", .fmt = "d-ng|n-ng"},
 	{.name = "ima-sig", .fmt = "d-ng|n-ng|sig"},
-	{.name = "ima-identity",.fmt = "n-ng|id|d-ng",
+	{.name = "ima-identity",.fmt = "n-ng|actor|subject",
 	 .genhash = ima_identity_genhash},
 	{.name = "", .fmt = ""},	/* placeholder for a custom format */
 };
@@ -40,7 +40,9 @@ static struct ima_template_field supported_fields[] = {
 	 .field_show = ima_show_template_string},
 	{.field_id = "sig", .field_init = ima_eventsig_init,
 	 .field_show = ima_show_template_sig},
-	{.field_id = "id",.field_init = ima_eventid_init,
+	{.field_id = "actor",.field_init = ima_event_actor_init,
+	 .field_show = ima_show_template_digest},
+	{.field_id = "subject",.field_init = ima_event_subject_init,
 	 .field_show = ima_show_template_digest}
 };
 
diff --git a/security/integrity/ima/ima_template_lib.c b/security/integrity/ima/ima_template_lib.c
index dc2492dd6658..f4a8c237acc0 100644
--- a/security/integrity/ima/ima_template_lib.c
+++ b/security/integrity/ima/ima_template_lib.c
@@ -330,10 +330,10 @@ out:
 }
 
 /*
- *  ima_eventid_init - Include identity of event in template.
+ *  ima_event_actor_init - Generate identity of actor involved in event.
  */
-int ima_eventid_init(struct ima_event_data *event_data,
-		     struct ima_field_data *field_data)
+int ima_event_actor_init(struct ima_event_data *event_data,
+			 struct ima_field_data *field_data)
 {
 	char actor[WP256_DIGEST_SIZE];
 	enum data_formats fmt = DATA_FMT_HEX;
@@ -347,3 +347,20 @@ int ima_eventid_init(struct ima_event_data *event_data,
 					   fmt, field_data);
 	return rc;
 }
+
+/*
+ *  ima_event_subject_init - Generate identity of subjdct involved in event.
+ */
+int ima_event_subject_init(struct ima_event_data *event_data,
+			   struct ima_field_data *field_data)
+{
+	int rc = 0;
+
+	if (event_data->file) {
+		pr_info("[%s]: inode=%s\n", __func__,
+			event_data->file->f_path.dentry->d_name.name);
+	}
+
+	rc = ima_eventdigest_ng_init(event_data, field_data);
+	return rc;
+}
diff --git a/security/integrity/ima/ima_template_lib.h b/security/integrity/ima/ima_template_lib.h
index 81c5453bee53..a08e6456b556 100644
--- a/security/integrity/ima/ima_template_lib.h
+++ b/security/integrity/ima/ima_template_lib.h
@@ -38,8 +38,10 @@ int ima_eventname_ng_init(struct ima_event_data *event_data,
 			  struct ima_field_data *field_data);
 int ima_eventsig_init(struct ima_event_data *event_data,
 		      struct ima_field_data *field_data);
-int ima_eventid_init(struct ima_event_data *event_data,
-		      struct ima_field_data *field_data);
+int ima_event_actor_init(struct ima_event_data *event_data,
+			struct ima_field_data *field_data);
+int ima_event_subject_init(struct ima_event_data *event_data,
+			struct ima_field_data *field_data);
 int ima_hostid_init(struct integrity_iint_cache *iint, struct file *file,
 		    const unsigned char *filename,
 		    struct evm_ima_xattr_data *xattr_value, int xattr_len,
-- 
2.11.0

